Docs and migration for 1.0.0 (#225)
Azure/azure-kusto-go: MIGRATION.md

# Migration Guide for Azure Data Explorer (Kusto) Go SDK

Welcome to the migration guide for the Azure Data Explorer (Kusto) Go SDK. This guide is designed to assist you in migrating your application from using the beta version of the SDK (`github.com/Azure/azure-kusto-go`) to the new, stable release, version `1.0.0`.

The release of version `1.0.0` introduces breaking changes, which include changes in dependencies, package arrangement, querying methods, and management commands. Following these steps carefully will ensure a smooth transition to the new version.

## 1. Changes in Dependencies and go.mod File

> To migrate from the old SDK to the new one, you'll need to update your `go.mod` file's dependencies.

Reviewer: ```suggestion
To migrate to version 1.x from older versions, you'll need to update your `go.mod` file's dependencies.
```

> **Old SDK:**

Reviewer 2: why needed ? can just put the new ..

Reviewer: Very much, this being a migration guide.

```bash
go get github.com/Azure/azure-kusto-go
```

**New SDK:**
- For data operations (`github.com/Azure/azure-kusto-go/azkustodata`):
```bash
go get github.com/Azure/azure-kusto-go/azkustodata
```

- For ingestion operations (`github.com/Azure/azure-kusto-go/azkustoingest`):
```bash
go get github.com/Azure/azure-kusto-go/azkustoingest
```

> Manually update your go.mod file by replacing `github.com/Azure/azure-kusto-go` with the specific package(s) you need.

Reviewer 2: Alternatively ?

## 2. Changes in Package Arrangement

The SDK is now split into two separate packages:
- **azkustodata:** For querying and managing Azure Data Explorer clusters.
- **azkustoingest:** For ingesting data into Azure Data Explorer clusters.

### Importing the New Packages

Depending on your requirements, import one or both of these packages into your Go files.

**For Data Operations:**
```go
import (
    "github.com/Azure/azure-kusto-go/azkustodata"
)
```

**For Ingestion Operations:**
```go
import (
    "github.com/Azure/azure-kusto-go/azkustoingest"
)
```

### Using the New Packages

> Update your code to use the new packages.

Reviewer 2: ```suggestion
Update your code with the new package API
```

Author: I don't think this sentence is better, the point is the packages themselves are new

For exmaple:

Old SDK:
```go

> kustoConnectionStringBuilder := kusto.NewConnectionStringBuilder(endpoint)

Reviewer: This is just a KCSB.
Do you want to show the client?

Reviewer: Also, is there value in showing these segmented line cases or do we want to focus on scenario examples (query, query with params, management command, ingest from , stream from, ... )

Author: The idea is to demonstrate the new package name.

I wanted to give an overview of things they should change if they're migrating.
For more detailed examples they have the README and example files

```

New SDK (For Data Operations):
```go
kustoConnectionStringBuilder := azkustodata.NewConnectionStringBuilder(endpoint)
```

Same for ingestion operations:

> Old SDK Ingestion Client Creation:

Reviewer 2: ```suggestion
Old SDK Ingestion Client Creation(Queued Example):
```

```go
in, err := ingest.New(kustoClient, "database", "table")
```

New SDK Ingestion Client Creation (Queued Example):
```go
in, err := azkustoingest.New(kustoConnectionString)
```

## 3. Building Queries

> The new SDK introduces a new way to build queries.:

Reviewer 2: ```suggestion
The new SDK introduces a new way to build queries:
```

The old SDK used a `kusto.NewStmt` method to build queries:
```go
    query := kusto.NewStmt("systemNodes | project CollectionTime, NodeId")
```

> For the new SDK, use the `azkustodata/kql` package to build queries, which has a type-safe query builder:

Reviewer: is there an example of query parameters somewhere?

Author: Yes in the README.md

Should it also be in the migration?

Reviewer: only if there's a change to the class names or usage

        .AddLiteral(" and ")
        .AddLiteral("NodeId == ").AddInt(value) // outputs ['system nodes'] | where CollectionTime == datetime(2020-03-04T14:05:01.3109965Z) and NodeId == int(1)
```

## 4. Querying Data

The new SDK introduces a new way to query data. 

The old SDK used the `Query` method to query data:
```go
import github.com/Azure/azure-kusto-go/kusto/data/table

// Query our database table "systemNodes" for the CollectionTimes and the NodeIds.
iter, err := client.Query(ctx, "database", query)
if err != nil {
	panic("add error handling")
}
defer iter.Stop()

// .Do() will call the function for every row in the table.
err = iter.DoOnRowOrError(
    func(row *table.Row, e *kustoErrors.Error) error {
        if e != nil {
            return e
        }
        if row.Replace {
            fmt.Println("---") // Replace flag indicates that the query result should be cleared and replaced with this row
        }
        fmt.Println(row) // As a convenience, printing a *table.Row will output csv
        return nil
	},
)
if err != nil {
	panic("add error handling")
}
```

The new SDK returns a dataset object, with the ability to handle different tables as they come:
```go

> dataset, err := client.QueryIterative(ctx, "database", query)

Reviewer 2: So we want iterative to be the default user behavior ?

Reviewer: I'd show a non-iterative and an iterative example in that order

Author: Better?

Reviewer: Yes

if err != nil {
    panic("error handling")
}
defer dataset.Close()

for tableResult := range dataset.Tables() {
    if tableResult.Err() != nil {
        panic("table error handling")
    }

>     for rowResult := range primaryResult.Table().Rows() {

Reviewer 2: primaryResult not set

        if rowResult.Err() != nil {
            panic("row error handling")
        }
        fmt.Println(rowResult.Row()) // Process each row
    }
}
```

You can also use the `Query` method to get all the results at once:
```go

> dataset, err := client.Query(ctx, "database", query)

Reviewer 2: add here how to get the primary resultsl

Author: Rewrote

  row.StringByIndex(0) // Get the value of the first column as a string
```

Or get the table as a slice of structs:
```go
import github.com/Azure/azure-kusto-go/azkustodata/query

  events, err := query.ToStructs[Event]()
  if err != nil {
    panic("error handling")
  }
    for _, event := range events {
        fmt.Println(event) // Process each event
    }
```

Management commands are now called using the `Mgmt` method on the client, and have an identical api to `Query`:
```go
  dataset, err := client.Mgmt(ctx, "database", query)
  if err != nil {
    panic("error handling")
  }
  primaryResult := dataset.Tables()[0]
  for _, row := range primaryResult.Rows() {
    fmt.Println(row) // Process each row
  }
```

## 5. Ingesting Data

The Ingestion API stayed the same, only using the new package:
```go
import github.com/Azure/azure-kusto-go/azkustoingest

kcsb := azkustodata.NewConnectionStringBuilder(`endpoint`).WithAadAppKey("clientID", "clientSecret", "tenentID")
ingestor, err := azkustoingest.New(kcsb, azkustoingest.WithDefaultDatabase("database"), azkustoingest.WithDefaultTable("table"))

if err != nil {
    // Handle error
}

> defer ingestor.Close() // Always close the ingestor when done.

Reviewer 2: shouldnt we close if err != nil ?

Author: If err != nil then the dataset is nil
